homeassistant/components/xiaomi_miio/switch.py

```diff
@@ -251,12 +342,150 @@ async def async_service_handler(service):
+    """Representation of a Xiaomi Plug Generic."""
+
+    def __init__(self, name, device, entry, unique_id, switch, coordinator):
+        """Initialize the plug switch."""
+        super().__init__(name, device, entry, unique_id, coordinator)
+
+        self._attr_icon = switch.icon
+        self._controller = switch
+        self._attr_is_on = self._extract_value_from_attribute(
+            self.coordinator.data, self._controller.short_name
+        )
+
+    @callback
+    def _handle_coordinator_update(self):
+        """Fetch state from the device."""
+        # On state change the device doesn't provide the new state immediately.
+        self._attr_is_on = self._extract_value_from_attribute(
+            self.coordinator.data, self._controller.short_name
+        )
+        self.async_write_ha_state()
+
+    @property
+    def available(self):
+        """Return true when state is known."""
+        if (
+            super().available
+            and not self.coordinator.data.is_on
+            and not self._controller.available_with_device_off
+        ):
+            return False
+        return super().available
+
+    @staticmethod
+    def _extract_value_from_attribute(state, attribute):
+        value = getattr(state, attribute)
+        if isinstance(value, Enum):
+            return value.value
+
+        return value
+
+    async def async_turn_on(self, **kwargs) -> None:
+        """Turn on an option of the miio device."""
+        method = getattr(self, SERVICE_TO_METHOD[self._controller.service]["method_on"])
+        if await method():
+            # Write state back to avoid switch flips with a slow response
+            self._attr_is_on = True
+            self.async_write_ha_state()
+
+    async def async_turn_off(self, **kwargs) -> None:
+        """Turn off an option of the miio device."""
+        method = getattr(
+            self, SERVICE_TO_METHOD[self._controller.service]["method_off"]
+        )
+        if await method():
+            # Write state back to avoid switch flips with a slow response
+            self._attr_is_on = False
+            self.async_write_ha_state()
+
+    async def async_set_buzzer_on(self) -> bool:
+        """Turn the buzzer on."""
+        return await self._try_command(
+            "Turning the buzzer of the miio device on failed.",
+            self._device.set_buzzer,
+            True,
+        )
+
+    async def async_set_buzzer_off(self) -> bool:
+        """Turn the buzzer off."""
+        return await self._try_command(
+            "Turning the buzzer of the miio device off failed.",
+            self._device.set_buzzer,
+            False,
+        )
+
+    async def async_set_child_lock_on(self) -> bool:
+        """Turn the child lock on."""
+        return await self._try_command(
+            "Turning the child lock of the miio device on failed.",
+            self._device.set_child_lock,
+            True,
+        )
```

Comment:
There should be no separate setters for these, but all of it should be generalized. Different devices have different actions, but there should be no need to duplicate this (almost) very same code for each and every togglable function.

Reply:
Just to give you some perspective: executing `ag 'def set_' miio/|wc -l` for the backend library yields 282 setters. Plenty of those belong to some specific devices that may be irrelevant (or are to be implemented by some other functions besides these specific service calls) but even then having a separate function for turning each property on as well as off makes an unsustainable overhead in the long run.

Reply:
I cannot commit to this branch directly, all PR's in this branch were already approved. Please create a new PR against the rewok branch to request the changes you want.

Reply:
> 
> 
> Just to give you some perspective: executing `ag 'def set_' miio/|wc -l` for the backend library yields 282 setters. Plenty of those belong to some specific devices that may be irrelevant (or are to be implemented by some other functions besides these specific service calls) but even then having a separate function for turning each property on as well as off makes an unsustainable overhead in the long run.

I think we can reduce the number of specific setters by making the code more abstract, so e.g. moving more to the SERVICE_TO_METHOD part and using getattr(). For now there seems no need for this yet. I even took away some abstraction in the select and number platform, because there was only sevice.